### ISSUE TITLE
Layers built during `call` are not tracked in complex model

### ISSUE THREAD

**Issue by jbischof** [keras-team/keras-core#707]
I've created a layer which creates a sublayer during `build` which is not built itself until `call`. Unfortunately, when used in a complex model, this extra layer and its weights are not tracked. Even adding 

```python
self.residual_projection.build(input_shape)
```

in the `build` does not seem to help, nor is it needed in the simple model.

```python
class ResnetBlock(keras.layers.Layer):
    def __init__(self, output_dim, **kwargs):
        super().__init__(**kwargs)
        self.output_dim = output_dim
        self.norm1 = keras.layers.GroupNormalization(epsilon=1e-5)
        self.conv1 = PaddedConv2D(output_dim, 3, padding=1)
        self.norm2 = keras.layers.GroupNormalization(epsilon=1e-5)
        self.conv2 = PaddedConv2D(output_dim, 3, padding=1)

    def build(self, input_shape):
        if input_shape[-1] != self.output_dim:
            self.residual_projection = PaddedConv2D(self.output_dim, 1)
        else:
            self.residual_projection = lambda x: x

    def call(self, inputs):
        x = self.conv1(keras.activations.swish(self.norm1(inputs)))
        x = self.conv2(keras.activations.swish(self.norm2(x)))
        return x + self.residual_projection(inputs)
```

The layer should have 10 weights if the input shape and output shape differ, but in the complex model it seems stuck at 8. Using `tf.keras` with the complex model, it has 10.

Complex model [gist](https://colab.research.google.com/gist/jbischof/996b5bab7c36f79d10ca111a7c86a9be/imageencoder-weights-issue-keras-core-tf-inlined.ipynb)
Simple model [gist](https://colab.research.google.com/gist/jbischof/e61541215a03b50b3e1349174278dc2d/resnetblock-build-issue-keras-core.ipynb)

**Comment by fchollet** [comment]
Thanks for the report. I looked into it, and it's a very subtle bug. I haven't fixed it yet. In short the problem is that the residual connection layer isn't tracked by its parent layer. It isn't tracked because tracking is disabled when it gets assigned. It's disabled because the assign happens, overall, inside the `__init__` of a `Sequential` model, which is exempt from tracking. So it's pretty complicated.

The solution is going to be to better scope the tracking exemption is Sequential so that it doesn't affect children layers.

**Comment by fchollet** [comment]
This should now be fixed -- please try it with the latest version at HEAD.

**Comment by jbischof** [comment]
Thanks for the quick fix! Unfortunately I hit #716 when trying to rebase at HEAD.

**Comment by jbischof** [comment]
Thanks, we're up to 10 layers now! See updated [gist](https://colab.research.google.com/gist/jbischof/996b5bab7c36f79d10ca111a7c86a9be/imageencoder-weights-issue-keras-core-tf-inlined.ipynb).

One note: the layer names no longer match `tf.keras`. We seem to be incrementing the index twice for every new layer (see truncated summary below). If this isn't a concern, feel free to close, but this would affect any `model.get_layer(name)` workflows.
```
 Layer (type)                    ┃ Output Shape              ┃    Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
│ padded_conv2d_1 (PaddedConv2D)  │ (None, None, None, 128)   │      3,584 │
├─────────────────────────────────┼───────────────────────────┼────────────┤
│ resnet_block_1 (ResnetBlock)    │ (None, None, None, 128)   │    295,680 │
├─────────────────────────────────┼───────────────────────────┼────────────┤
│ resnet_block_3 (ResnetBlock)    │ (None, None, None, 128)   │    295,680 │
├─────────────────────────────────┼───────────────────────────┼────────────┤
│ padded_conv2d_11 (PaddedConv2D) │ (None, None, None, 128)   │    147,584 │
├─────────────────────────────────┼───────────────────────────┼────────────┤
│ resnet_block_5 (ResnetBlock)    │ (None, None, None, 256)   │    919,040 │
├─────────────────────────────────┼───────────────────────────┼────────────┤
│ resnet_block_7 (ResnetBlock)    │ (None, None, None, 256)   │  1,181,184 │
├─────────────────────────────────┼───────────────────────────┼────────────┤
```
